api/src/main/java/org/apache/iceberg/types/Types.java

```diff
@@ -414,42 +414,71 @@ public int hashCode() {
 
   public static class NestedField implements Serializable {
     public static NestedField optional(int id, String name, Type type) {
-      return new NestedField(true, id, name, type, null);
+      return new NestedField(true, id, name, type, null, null);
     }
 
     public static NestedField optional(int id, String name, Type type, String doc) {
-      return new NestedField(true, id, name, type, doc);
+      return new NestedField(true, id, name, type, null, doc);
+    }
+
+    public static NestedField optional(int id, String name, Type type, Object defaultValue, String doc) {
+      return new NestedField(true, id, name, type, defaultValue, doc);
+    }
+
+    public static NestedField optionalWithDefault(int id, String name, Type type, Object defaultValue) {
+      return new NestedField(true, id, name, type, defaultValue, null);
     }
 
     public static NestedField required(int id, String name, Type type) {
-      return new NestedField(false, id, name, type, null);
+      return new NestedField(false, id, name, type, null, null);
     }
 
     public static NestedField required(int id, String name, Type type, String doc) {
-      return new NestedField(false, id, name, type, doc);
+      return new NestedField(false, id, name, type, null, doc);
+    }
+
+    public static NestedField required(int id, String name, Type type, Object defaultValue, String doc) {
+      validateDefaultValueForRequiredField(defaultValue);
+      return new NestedField(false, id, name, type, defaultValue, doc);
+    }
+
+    public static NestedField requiredWithDefault(int id, String name, Type type, Object defaultValue) {
+      validateDefaultValueForRequiredField(defaultValue);
+      return new NestedField(false, id, name, type, defaultValue, null);
     }
 
     public static NestedField of(int id, boolean isOptional, String name, Type type) {
-      return new NestedField(isOptional, id, name, type, null);
+      return new NestedField(isOptional, id, name, type, null, null);
     }
 
     public static NestedField of(int id, boolean isOptional, String name, Type type, String doc) {
-      return new NestedField(isOptional, id, name, type, doc);
+      return new NestedField(isOptional, id, name, type, null, doc);
     }
 
+    public static NestedField of(int id, boolean isOptional, String name, Type type, Object defaultValue, String doc) {
+      return new NestedField(isOptional, id, name, type, defaultValue, doc);
+    }
+
+    private static void validateDefaultValueForRequiredField(Object defaultValue) {
```

Comment:
Might be better to just move this check into the constructor of `NestedField`.

Reply:
makes sense

Reply:
so, we are using defaultValue = null to represent no-defaultValue. 
The validation was meant to avoid the case when the user calls an API that passes defaultValue and sets the defaultValue to null in error (to avoid behavior that might be unexpected by the user). However, we can say it is legal to set defaultValue to null, which is equivalent to calling and API with no defaultValue, and we leave it to the user's responsibility to user the right APIs and parameters values (with adding java docs regarding defaultValue to the corresponding APIs). Otherwise, the validation will have to take place in the relevant APIs, not the constructor (because at constructor level it is legal) OR, we will have to use a different mechanism to represent no default value. For example, adding a _hasDefault flag, which is not ideal (extra effort to API callers) or use Optional which is not recommended for parameters.